pppd/main.c

```diff
@@ -747,6 +750,108 @@ set_ifunit(int iskey)
     }
 }
 
+/*
+ * net-* scripts to be run come through here.
+ */
+static void main_script(char* script, int wait)
+{
+    char strspeed[32];
+    char *argv[6];
+
+    slprintf(strspeed, sizeof(strspeed), "%d", baud_rate);
+
+    argv[0] = script;
+    argv[1] = ifname;
+    argv[2] = devnam;
+    argv[3] = strspeed;
+    argv[4] = ipparam;
+    argv[5] = NULL;
+
+    run_program(script, argv, 0, NULL, NULL, wait);
+}
+
+/*
+ * set_ifup - called in order to set the ppp interface to up, if not already
+ * brought up.
+ *
+ * It's important that the sub-protocol not be a dynamically allocated string, ie,
+ * use set_ipup("IPCP") ...
+ */
+int
+set_ifup(const char* name)
```

Comment:
I am not a big fan of the new static variable up_protos[]. The protocols are already a part of the protocols[] (struct protoent). Seems like there should be a possibility of querying that structure to figure out if it is set in state up or not. 

Also, I would imagine that NET_PRE_UP would be executed like IP_PRE_UP which means all configured network protocols have marked themselves as done (either UP or FAIL), but interface is not yet in configured with IFF_UP yet). 

A script could possibly query the configuration of the interface being brought up to find the values it needs if it isn't passed into the net-pre-up script (as ip-pre-up would pass in the IP information, or the IPv6 information).

Reply:
array of ints vs array of pointers?  I was also wondering how to avoid up_protos, but alas, couldn't think of any.  Even considered a bitmask ... do you have a better suggestion?

Yes, NET_PRE_UP gets executed like IP_PRE_UP, but since we cannot determine the order in which protocols configure, we can make no guarantees w.r.t. configured status (some protocol - probably exactly one - may already be configured, but it could be either ipv6cp or ipcp or theoretically some other), only that the interface will not yet be in IFF_UP state, but about to transition into that.

Yes, a script can either query the details using say "ip" or by using the environment variables (as set by script_setenv calls scattered through pppd - but again, no guarantees can be made since probably only one of ipv6cp or ipcp will have completed at this stage).

Reply:
Eliminated the sys-* netif_set_up and netif_set_down functions in preference of exposing a single netif_set_state() from each.

Reply:
I think I would look at main.c and find how they processes the protocol structure in get_input(), but if you could add a int field in there that says state, and this new function would 
pseudo code:

for (protocol entry : protocols) { 
   if entry.id == PPP_EAP || entry.id == PPP_LCP || entry.id == .. ) // or, add a field phase in each protocol entry, i.e. mass update the files, then if (PHASE_NETWORK != entry.phase)
      continue;
   if !entry.up
      return -1
}

// all NETWORK protocols are UP here.
setifstate(unit, 1);

Generally, adding the extra information to protocols structure; could help readability of the application. PHASE_NETWORK would be up when all NETWORK protocols has been negotiated. 

Similarly, a flag to say the protocol is an AUTH protocol, you could chose to only pass the information to that protocol instead of checking EAP and LCP. See RFC1661.

Reply:
Trying to run NET_PRE_UP before all protocols has completed negotiation runs you afoul with the problem that not all information can be extracted. Though, trying to pass it to the script may be difficult, but a script can be made to query the state of the interface and extract the IPv6, IPv4 and MPLS information. If you execute the NET_PRE_UP before all network protocols have been negotiated then this information would be lacking at that point.

Reply:
OK, this makes sense to a degree and I can get behind that.  This would restore the behaviour for ip-pre-up too in that if ipv6cp completes before ipcp it would prevent pppd from bringing up the interface until ipcp too completed.  I can, to a degree, get behind that.

Our use-case frankly doesn't care about any of the *potentially* configured details, we merely need to know (info from pppd):

1.  Radius tags that was received as part of Access-Accept (radattr plugin) during authentication.
2.  Authenticated username (typically not in radattr data).
3.  ipparam string.

And I highly prefer the interface still being *down* in order to avoid possible NAT (conntrack) related problems.  I'd rather generate a ICMP(v6) no route to network than inject an incorrect conntrack entry into CT (TCP is OK normally, connection will timeout and typically retry, but again, rather ECONNREFUSED/ENETUNREACH than ETIMEOUT, UDP can be much more problematic, and we do see cases of that).

The NAT is an issue because unit numbers are typically dynamically allocated, and some ppp interfaces needs MASQUERADE, others do not ... so having a pre-set up NAT isn't always viable.  IPv6 doesn't need NAT, so not a problem there.

So for *our use case* this is good enough, but I'm happy to accommodate as wide as reasonably possible.  So one thing I have seen in the past (and this is probably a buggy peer) was IPV6CP that kept going back and forth indefinitely.  Given your suggestion, a case like that would completely prevent the interface from eventually coming up.

So here is my suggestion, and this breaks backwards compatibility:

Update ip-pre-up to run PRIOR to configuration of the interface, passing the details that WILL BE CONFIGURED as per current.  This will need to be made VERY CLEAR in upgrade documents (ChangeLog and/or Changes-2.5.0), as well as the man page.  We don't require ipv6-pre-up, but the same logic could then be applied for ipv6cp should a ipv6-pre-up get added at a later stage in that module, ie, run the script PRIOR to configuration, passing the "will be used" details to the script.

One *could* even (since we're breaking backwards compatiblity with such a change) even indicate that should this script exist, and exit non-zero, restart negotiation for IPCP (and IPV6CP).  But I don't think this is a good idea, should a script want that, just kill -HUP $PPPD_PID - however, that restarts negotiation for ALL protocols.

Reply:
@jkroonza 

I am coming from the perspective having implemented ip-up and ip6-up callbacks within a plugin, see 
* https://gitlab.gnome.org/GNOME/network-manager-sstp/-/commit/5e3e3e9357454c0c573c1b62ee19e19c6b724fa5, and generally:
*https://gitlab.gnome.org/GNOME/network-manager-sstp/-/blob/master/src/nm-sstp-pppd-plugin.c

While the network-manager stuff doesn't necessarily use the scripts, it hooks into the notifier sub-systems from within a plugin. Similarly, I would hope that a net-pre-up would also dispatch a notifier saying all protocols up, "I am about to set IFF_UP flag", because at that point the plugin can assemble the configuration and send it off to Network Manager in one go instead of tracking the details of the ipcp or ip6cp protocol structures and only move forward when both are configured (or one configured, the other disabled, vice versa). 

I would imagine that the current implementation of ip-pre-up satisfied most use-cases back in 2005, but later the environment has changed, i.e. introduction and more frequent use of IP6CP, and in the future possible others.

I'd like for the functionality to add another "state" such that 
* All "network" protocols (see RFC1661) are in state up, but before interface is configured in state UP (IFF_UP)
* A notifier is dispatched when net-pre-up is issued
* Execute the net-pre-up, that allows for what your needs are
** Rename interface
** Configure NAT, MASQUERADE, ec
** Doesn't provide necessarily all parameters, but could set environment variables such as an accumulation of "IPV6_ADDR", IPV4_ADDR, IPV4_DNS, IPV4.."
** Still provides the basics arguments to the scripts such as speed, device name, etc.
* Configure interface with IFF_UP
* Anything else?

I'd much rather "design" the feature knowing the outcome instead of trying to get things done in a hurry. It should be an improved version of the ip-pre-up that encompasses all protocols instead of just IP.  A follow up change could be to completely remove the ip-pre-up functionality breaking backwards compatibility ensuring that users update their scripts and also where they read the variables from (not argument list, but environment, etc). 

Does this make sense?

Reply:
Also, the current protocol structure doesn't state if a protocol was intended to run during a particular phase, that could possibly be added in a separate change to ensure once you iterate through the structure don't have to e.g. compare it against a set of hard-coded protocol numbers (e.g. see get_input() in main.c).

This would be helpful here too where you iterate through the protocols and would only set the interface in state up when all the protocols has come up. Unfortunately, this probably require it to expose yet another callback to the protocol code to say: yeah you are up and scripts like ip-up and ip6-up gets executed (after interface has been configured in state up (IFF_UP)).

Reply:
@enaess thinking on this I agree this would be nice.

However, and this is a big issue, both ip-pre-up and ip-up is called from the same function in ipcp.c - this is a sequential step.  So this would require a fairly major redesign.

I can certainly add a notify() and/or net_{pre_,}up_hook() mechanism for you, that's quite trivial.

Do we know of many systems using ip-pre-up?  My recommendation would be that protocol-specific pre-ups remain, but that they be moved *prior* to interface configuration.  This does of course imply that net-pre-up will likely be called *after* one protocol-specific pre-up, which I think is not ideal.  Unless all protocols adheres to:

1.  Ensure iface is UP.
2.  Invoke protocol-specific pre-up (currently only ip, not ipv6), providing "will be used" details.
3.  Configure interface for protocol.
4.  Invoke up (currently ip and ipv6).

Alternatively (very major rework referred to above) we can somehow permit all protocols to do something like the following:

1.  Upon starting pppd, prior to proceeding indicate that the protocol (if to be negotiated) needs to finish negotiation prior to interface up.
2.  If negotiation fails remove from the "desired" set.
3.  If negotiation succeeds, configure the interface as required.
4.  Invoke pre-up. (This can potentially move to pre-up callback in 5 in order to only invoke it *after* net-pre-up I reckon).
5.  Call if_setup(), additionally providing two optional call-backs.  1 to invoke when ready to bring interface up, just prior to actually doing so.  1 to invoke just after bringing the interface up.

All functionality in the current _up functions after the call to if_setup would thus move into the latter callback.  I don't see a specific need to the first callback, but I'd rather someone look at it rather than for it.

One problem I possibly foresee is ondemand ... but I'd have to take a much closer look.  As I see it there would then be four net-* scripts:

1.  net-init - as per current.
2.  net-pre-neg (need a better name, net-ready came to mind but that implies to iface is ready to bring up, which is not the case) - invoked post authentication but prior to initiating IPCP and IPV6CP (and others).
3.  net-pre-up - invoked just prior to the pre-up callbacks (such that it's possible to get ordering net-pre-up followed by proto-pre-up).
4.  net-down (as current).

Would this be more suitable?  What if we WANT to bring up the interface when the first protocol is ready instead of when ALL protocols are ready?  My opinion is that the current setup kinda works for me in that once the first protocol is good to go the interface can come up, so if IPv6 goes back and forth, failing to negotiate, why should it delay IPv4 or the other way around?

Reply:
@enaess could you please provide some feedback here?

Reply:
I think the general outline of the above seems okay. However, I would want to solicit feedback from @paulusmack to make sure he is okay with a change like that and what you are trying to do. The other part of this is how "risky" is a change like that, meaning will it break or become fragile, e.g. is ip-pre-up / net-pre-up deterministic in all cases like ip is up, but ip6 failed negotiation, the test-coverage, etc.

Reply:
@paulusmack @enaess may I please request some direction here?